pyVHDLParser/Blocks/Reference/Context.py

```diff
@@ -27,139 +27,329 @@
-		}
+@export
+class StartBlock(Block):
+	"""
+	This block is used to disambiguate between context declarations and context references.
+
+	The states in this class only check whether the token stream represents a declaration or a reference.
+	They neither change :class:`Token`s nor do they create new :class:`Block`s.
+	After a decision can be made, the unhandled tokens (everything between :attr:`parserState.TokenMarker` and the current token)
+	are parsed a second time by either :attr:`ReferenceStartBlock.fromTokenMarker` or :attr:`DeclarationStartBlock.fromTokenMarker`.
+	"""
 	@classmethod
 	def stateContextKeyword(cls, parserState: TokenToBlockParser):
+		cls.stateWhitespace1(parserState)
+
+	@classmethod
+	def stateWhitespace1(cls, parserState: TokenToBlockParser):
 		token = parserState.Token
-		if isinstance(token, WhitespaceToken):
-			parserState.NewToken =    BoundaryToken(fromExistingToken=token)
-			parserState.NextState =   cls.stateWhitespace1
-			return
-		elif isinstance(token, (LinebreakToken, CommentToken)):
-			block =                   LinebreakBlock if isinstance(token, LinebreakToken) else CommentBlock
-			parserState.NewBlock =    cls(parserState.LastBlock, parserState.TokenMarker, endToken=token.PreviousToken, multiPart=True)
-			_ =                       block(parserState.NewBlock, token)
-			parserState.TokenMarker = None
-			parserState.NextState =   cls.stateWhitespace1
+
+		if _isNonCodeToken(token):
+			parserState.NextState = cls.stateContextName
 			return
 
 		raise BlockParserException("Expected whitespace after keyword CONTEXT.", token)
 
 	@classmethod
-	def stateWhitespace1(cls, parserState: TokenToBlockParser):
+	def stateContextName(cls, parserState: TokenToBlockParser):
 		token = parserState.Token
+
+		if _isNonCodeToken(token):
+			return
+
+		if isinstance(token, (WordToken, ExtendedIdentifier)):
+			parserState.NextState =     cls.stateWhitespaceOrSemicolonOrDot
+			return
+
+		raise BlockParserException("Expected context name (identifier).", token)
+
+	@classmethod
+	def stateWhitespaceOrSemicolonOrDot(cls, parserState: TokenToBlockParser):
+		token = parserState.Token
+
+		if _isNonCodeToken(token):
+			parserState.NextState =        cls.stateIsOrSemicolonOrDot
+			return
+
+		if isinstance(token, CharacterToken) and token == ";":
+			parserState.ReparseFromTokenMarker(ReferenceStartBlock.stateFromStartBlock)
+			parserState.Pop()
+			return
+		elif isinstance(token, CharacterToken) and token == ".":
+			parserState.ReparseFromTokenMarker(ReferenceStartBlock.stateFromStartBlock)
+			return
+
+		raise BlockParserException("Expected whitespace after context name (identifier).", token)
+
+	@classmethod
+	def stateIsOrSemicolonOrDot(cls, parserState: TokenToBlockParser):
+		token = parserState.Token
+
+		if parserState.HandleNonCodeTokens(cls):
+			return
+
+		if isinstance(token, WordToken) and (token == "is"):
+			parserState.ReparseFromTokenMarker(DeclarationStartBlock.stateFromStartBlock)

```

Comment:
The state machine decided, that the block can only represent a context declaration. `ReparseFromTokenMarker` only needs to know in which state the second parsing should be started. The first token passed is the `TokenMarker` itself.